### ISSUE TITLE
Using multiple "flutter apps" inside a page using Element Embedding

### ISSUE THREAD

**Issue by eduardo-pellenz** [flutter/flutter#121374]
## Use case

I'm currently writing a chart library for the company that I work for. After some research, we didn't find a good package that have the features that we want. So our decision was to write our own.

We have both an web app and app (React and Flutter). And with the `Element Embedding` feature we want to reuse this package for both platforms.

After some tests using the `master` branch, there was two things that I found that would be really cool to add/improve (don't actually know if it's on your roadmap already)

### I wasn't able to create two instances of the app. Is there or will be a way to create multiple instances of the app in a page? This is the code that I have used.

```dart
const initChart = (targetSelector) => {
  let target = document.querySelector(targetSelector);
  _flutter.loader.loadEntrypoint({
    onEntrypointLoaded: async function (engineInitializer) {
      let appRunner = await engineInitializer.initializeEngine({
        hostElement: target,
      });
      await appRunner.runApp();
    },
  });
}

window.addEventListener("load", function (ev) {
  initChart('#flutter_target');
  initChart('#flutter_target_2');
});
```

With this code, the first app is stopped and the second one works.

### Will be there a way to export single widgets instead of having to export the whole app?

Using the current way, it's possible to achieve what I want using a wrapper around the `_flutter.loader.loadEntrypoint` and calling the state methods to update the state. But will be there a better way to do that?

**Comment by huycozy** [comment]
Hi @eduardo-pellenz 
There is a related document about this ([Astral Projections for Flutter Web](https://docs.google.com/document/d/1pvH-J8opXsjntTpFf1bbsbU8N1vFd8tPKvBgcNn8UKQ)). And it seems the multi-screen Flutter web app hasn't been supported yet.

Labeling the issue as a proposal for more infights. 

cc @ditman for thoughts.

**Comment by ditman** [comment]
**For now**:

> Is there or will be a way to create multiple instances of the app in a page?

**We cannot have multiple instances of a Flutter app in the same page.** Flutter web is using some window globals to keep its state that can get messed up if multiple flutter engines write on them. This needs fixing by the flutter engine. I don't have an ETA for this.

> Will be there a way to export single widgets instead of having to export the whole app?

Not really. I don't see a future where you're going to be able to do: `import flutter/Scaffolding from main.dart.js` and use that in your JS. Similarly to how you wouldn't import a single ReactJS component outside of a ReactJS app?

The [next steps](https://github.com/flutter/flutter/issues/118481) for this are to support multiple "screens" in Flutter web as different elements on the page, so I guess you could have some divs that "only render one widget" as a proxy for exporting single widgets, or running multiple instances of the app.

(Something very similar to this was discussed in our [Discord recently, check it out!](https://discord.com/channels/608014603317936148/1072783325757976606))

@eduardo-pellenz I wanted to thank you for creating this issue. We're attempting to build this in a way where *users* are more involved, and all these doubts, feature requests and "I think this'd work X way but it does Y" posts help very much! Keep it up!

**Comment by jacobsimionato** [comment]
The inability to have multiple separate Flutter instances coexist in the same app is a potential problem when using Flutter to implement a shared library on the web, as there could be other Flutter-based libraries that already exist in the page. This affects potential internal Google clients too.

**Comment by jehuamanna** [comment]
I have a need to integrate multiple web apps written in Flutter into a final single web app. I am successful in loading assets, main.dart.js files, and routing the web app using single-spa. The only concern is the error message due to loading multiple main.dart.js files.


This is a route-based micro-frontend use case. I hope I get some pointers to move forward from dart-web team.  I don't know how I can isolate these main.dart.js scripts. I renamed them( for the sake of it) and still, they are causing issues.

I also thought about the iframe method. But still not been successful.

**Comment by albertolina** [comment]
Hello, I have the same need. Any news about this issue?

**Comment by ditman** [comment]
@albertolina  @jehuamanna  The work thas has been happening recently in this space has to do with adding multiple "views" to a single flutter app. It'd probably be sufficient for @eduardo-pellenz's original request.

One flutter app can drive as many instances of a widget in a page. For example:

* Passing arbitrary data from JS into Flutter: https://dit-multiview-tests.web.app
* Rendering many platform views (Google Maps): https://dit-maps-tests.web.app

This is what's in progress, but still in development. This ability makes some Flutter web shortcomings much more apparent, like scrolling events being "stolen" by flutter and things like that. Still working on this!

**Comment by albertolina** [comment]
Thank you! Amazing, can't wait to use these!

**Comment by albertolina** [comment]
Hi @ditman,

Our team is currently evaluating the best approach to integrate a segment of our main application into our website. We're weighing the options between developing a lightweight version using JavaScript/React and waiting for a specific Flutter feature currently in development. Given the potential for reduced development and maintenance effort, we're highly interested in the Flutter option.

Could you provide any information on whether there's a possibility to test this Flutter feature through a beta version or a specific branch? We understand the risks associated with using beta software, but we're in a testing phase with this project and are willing to explore this route if it aligns with our timeline and requirements.

Thanks!

**Comment by ditman** [comment]
@albertolina all the changes are available in the `master` branch of flutter. We're aiming to have a first MVP with documentation for the next stable release.

**Comment by matthew-carroll** [comment]
I'd like to add my use-case to this. I'm building out a number of Flutter education websites. I'd like to include many Flutter samples throughout the websites. This includes individual pages with multiple Flutter samples.

I was hoping that the work that made Flutter embeddable in HTML elements would make this trivial. It was unfortunate to see that only a single Flutter experience can be presented within a given page at a time.

**Comment by dkbast** [comment]
I have the exact same usecase as @matthew-carroll - using iframes & dartpad at the moment but I'm definitely looking forward to using element embedding instead. The multi counter example looks nice - any chance you could link to the source of that?

**Comment by ditman** [comment]
@dkbast there's a bunch of samples here, but the API is evolving so don't take them too seriously: https://github.com/ditman/flutter-experimental/tree/main/multi_counter

**Comment by ditman** [comment]
> I'd like to add my use-case to this. I'm building out a number of Flutter education websites. I'd like to include many Flutter samples throughout the websites. This includes individual pages with multiple Flutter samples.
> 
> I was hoping that the work that made Flutter embeddable in HTML elements would make this trivial. It was unfortunate to see that only a single Flutter experience can be presented within a given page at a time.

@matthew-carroll Can you elaborate a little bit more? Why do you need more than one flutter app on your page? You should be able to render whatever widget you want on each view from a single flutter app (I think)?

**Comment by matthew-carroll** [comment]
> Why do you need more than one flutter app on your page? You should be able to render whatever widget you want on each view from a single flutter app (I think)?

I don't care whether it's multiple engines or a single engine with multiple views, but I need to display Flutter stuff in multiple DOM elements. Moreover, I obviously also need to have full control over what's displayed in each Flutter DOM embedding. If the way that's implemented is multiple views in one engine, that's fine.

**Comment by matthew-carroll** [comment]
> add/remove views to a running Flutter Web app from JS

Is there a declarative version, too? Again, in my case, I'm going to have a large number of simplistic views, so writing JS to wire all of them up would be disproportionately onerous as compared to the simplicity of what I'm trying to display.

> When adding a view you can pass "initial data" from JS and "constraints"

Can we get a sane default for the `constraints`? For example, I'm likely to set the desired width and height of my DOM element through CSS, and the obvious default constraints would be "make yourself the same size as the DOM element you inhabit". I'd prefer to not need to repeat that on every embedded widget.

On a related note, is the reverse direction being handled, too? 

Imagine that I choose to display a code syntax highlighter on my web page that's rendered by Flutter. I put a long code snippet in the widget, so initially the code syntax highlighter choose to cut off the code beyond, let's say, line 10, and add a shadow to the bottom to indicate there's more code. When the user taps on the code snippet, the widget wants to fully extend itself to show all the code. What I'm describing here is a specific example of intrinsic height.

Will there be a concise and declarative way to tell the embedded Flutter view to size its DOM element to match the Flutter view's intrinsic height?

**Comment by ditman** [comment]
> Is there a declarative version, too?

Initially, **no**, but if you want something more declarative you can hide the complexity of adding/removing views behind a `<custom-element>` (whether standard HTML, Angular, React or whatever element library you're using on your web app. We did something like that [here](https://github.com/flutter/samples/tree/main/web_embedding) (see `ng-flutter` for a FlutterApp Angular element)).

> the obvious default constraints would be "make yourself the same size as the DOM element you inhabit"

Of course, this is the current behavior of embedding, and the default for the next version. Each constraint value will default to the current size of the container element; in Flutter speak: "Tight to the htmlElement".

> [...] initially the code syntax highlighter choose to cut off the code beyond, let's say, line 10, and add a shadow to the bottom to indicate there's more code. When the user taps on the code snippet, [...] show all the code. Will there be a concise and declarative way to tell the embedded Flutter view to size its DOM element to match the Flutter view's intrinsic height?

I think this is achievable with a `minHeight` constraint to the ~height of your 10 lines (or zero), and the `maxHeight` constraint to `Infinity`. That'll let your Flutter web widget grow unbounded vertically (you can also set a specific max, so you can end up with expanded code + scroll, for example).

Note that your Flutter container shouldn't have certain CSS that would defeat this behavior even in standard HTML. For example: a hardcoded `height`, `max-height` with `overflow: hidden` that would prevent anything with intrinsic sizing from expanding its parent container past the defined height.

Flutter won't fight against your CSS when applying its intrinsic sizing (similar to an `img` element, I guess)

**Comment by matthew-carroll** [comment]
My point about the intrinsic height wasn't about how to accomplish that in Flutter. I know how to do that. My point was that when the Flutter widget changes its intrinsic height from shorter to taller, will there be an affordance provided by Flutter to automatically change the height of the DOM element to match the new intrinsic height of the Flutter widget (if desired)?

**Comment by ditman** [comment]
> will there be an affordance provided by Flutter to automatically change the height of the DOM element to match the new intrinsic height of the Flutter widget (if desired)?

@matthew-carroll like this? https://dit-multiview-resize.web.app

There's no explicit affordance. Flutter will just set the dimensions of its own container by measuring the available space and combining that with the constraints that you pass to each view.

**Comment by matthew-carroll** [comment]
Yes that example is what I mean. Although that example doesn't change the intrinsic height, so it's possible that Flutter initializes to intrinsic size and then ignores changes to the intrinsic size.

But this goes back to my earlier question about explicit size.

Are you saying that Flutter automatically chooses between the explicit bounding size and the intrinsic size approach?

Case 1: I want my Flutter view to be 400px wide by 200px tall.

Case 2: I want my Flutter Element to be exactly the same width and height as the Flutter widget.

How does the developer switch between those modes of sizing?

**Comment by ditman** [comment]
@matthew-carroll I'm not sure I've tried that case. Flutter views can be intrinsically sized (with constraints) or sized by CSS and filling the available space. In the first version you won't be able to change the constraints of a flutter view after it's been created. (You can always detach a view and attach it again with new constraints, but that's probably not an "easy" affordance.)

**Comment by matthew-carroll** [comment]
@ditman in terms of informing the road map with use-cases, your description there suggests that I probably wouldn't be able to accomplish one of my desired goals. For that code snippet example, not only would I like for the DOM element to auto-resize based on the internal Flutter widget intrinsic height, but it would be even better if I could animate the height of my internal Flutter widget, so that the code snippet slides taller. That would be like a per-frame version of the same problem.

Please consider working that into the roadmap. As developer make tactical uses of embedded Flutter DOM elements, I'd bet that this affordance would be desired.

**Comment by ditman** [comment]
@matthew-carroll sure, let's see if the MVP fits your needs or not and work from there.

**Comment by khuntia** [comment]
@ditman For multiple independent flutter engines itself, is it not possible to compile them to wasm (now that it looks quite stable) and have them loaded as multiple wasm embeddings ? I am also curious to know what if one of them is wasm compiled while other is js, can they co-exist as well ?

**Comment by ditman** [comment]
> @ditman For multiple independent flutter engines itself, is it not possible to compile them to wasm (now that it looks quite stable) and have them loaded as multiple wasm embeddings ? I am also curious to know what if one of them is wasm compiled while other is js, can they co-exist as well ?

@khuntia no, the flutter bootstrap process and the web engine itself still use several global JS variables (on the JS `window`). Multiple instances would clobber each other at initialization and run-time. Multiple engines are definitely not supported.

**Comment by jehuamanna** [comment]
@ditman Thank you for providing the leads. With your demo and code by @goderbauer, I was able to get a working demo of running different applications in a single window. 
Here is the demo link. [Multiple application switcher](https://jehuamanna.github.io/mutiple-flutter-application-demo/)

It can be also made worked to run different application simultateous in a single window.

As you mentioned, there are few issues. The scroll gets highjacked. The navigation does not work intuitively. 

Let me know your thoughts, @ditman . Also, please guide me in next steps

**Comment by jehuamanna** [comment]
I am attaching the link to the blog as a write-up. [Link](https://jehuamanna.dev/2024/05/18/adding-multiple-flutter-applications-to-single-window-on-web/)

**Comment by ditman** [comment]
@jehuamanna, you're doing unsupported things and they can break in any unforeseeable way down the line. I'm sorry, but for now we're not working on enabling the use case of having multiple engines running in parallel in a page, [as stated above](https://github.com/flutter/flutter/issues/121374#issuecomment-1449221532).

**Comment by PraveenPushpkar** [comment]
Hi flutter team. I am working on a project for a client in my company where the requirement is to develop small individual components using flutter which can be dragged and dropped (using one of their CMS) to their normal html pages.
Since in a single page we can have just one flutter engine which currently allows just a single flutter component (main.dart.js file), we are stuck and the client might loose interest in using flutter for their future projects.

I request the flutter team to allow the use of multiple flutter components in a single page.
